lib/choice_type_creator.rb

```diff
@@ -0,0 +1,224 @@
+# frozen_string_literal: true
+
+require 'pry'
+require 'securerandom'
+
+module DataScript
+  # Class for making sure that every resource/profile that has a choice type (that's not
+  # already covered by Synthea) has every choice type represented in the output.
+  # For the moment, this only does Date/time related choice types, because they're easy to
+  # automatically generate.
+  class ChoiceTypeCreator
+    SWAP_COUNT = 1
+    MUST_SUPPORT_CHOICE_TYPES = {
+      FHIR::DiagnosticReport =>
+      {
+        prefix: 'effective',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-note',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-lab'
+        ]
+      },
+      FHIR::Immunization =>
+      {
+        prefix: 'occurrence',
+        suffixes: %w[DateTime String],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-immunization'
+        ]
+      },
+      FHIR::Observation =>
+      {
+        prefix: 'effective',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/StructureDefinition/resprate',
+          'http://hl7.org/fhir/StructureDefinition/heartrate',
+          'http://hl7.org/fhir/StructureDefinition/headcircum',
+          'http://hl7.org/fhir/StructureDefinition/bodyweight',
+          'http://hl7.org/fhir/StructureDefinition/bp',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-smokingstatus',
+          'http://hl7.org/fhir/us/core/StructureDefinition/pediatric-weight-for-height',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-observation-lab',
+          'http://hl7.org/fhir/us/core/StructureDefinition/pediatric-bmi-for-age',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-pulse-oximetry',
+          'http://hl7.org/fhir/StructureDefinition/bodyheight',
+          'http://hl7.org/fhir/StructureDefinition/bodytemp'
+        ]
+      },
+      FHIR::Procedure =>
+      {
+        prefix: 'performed',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure'
+        ]
+      }
+    }.freeze
+
+    # Goes through the set of bundles, and ensure we have at least SWAP_COUNT resources
+    # in each resource/profile pair listed above that conform to each choice type.
+    # @param bundles [Array] an array of FHIR::Bundle objects
+    # @return nil
+    def self.check_choice_types(bundles)
+      # Get the resources sorted by their resource type (fhir_models classname)
+      resource_map = get_resources_by_class(bundles)
```

Comment:
Rather than `get_resources_by_class`, why not just iterate over the profiles in `MUST_SUPPORT_CHOICE_TYPES` and have a `select` to grab all of the resources that fit a particular profile?

Reply:
Yeah, that would be another option. I think it might be a little harder to index into the `MUST_SUPPORT_CHOICE_TYPES` hash if the resources were collected by profile URL rather than class (though I guess you could use the `.class` of the object to get the key of the hash.

Reply:
Is it necessary to index into the hash? I was imagining something like:
```
MUST_SUPPORT_CHOICE_TYPES.each do |_, must_supports|
  must_supports[:profiles].each do |profile|
    all_the_bundled_resources
      .select { |resource| resource.meta.profile == profile }
      .each { |resource| process these resources }
  end
end
```

Reply:
I think this would be feasible. I'll investigate during the refactor for the comment below.